### PR TITLE
Pixly

### PR DESCRIPTION
**Pull request by zed9h**
First try at implementing the Pixly file format, just loading for now and I tested with only two files, but it could be useful to some people ([for instance this guy](https://trello.com/c/lyMh9uSc/13-import-psd-aseprite-gif-pcx)), so I am pushing right away.

### CODE REVIEW

**Comment by dacap** [comment]
Hi @zed9h, could this use the same png encoder/decoder to avoid duplicating the whole code? I see that it opens an XML file and a png file, so to load the png file you might try to use the png decoder (or calling `load_document()` to load the png file as a temporal `doc::Sprite`/`app::Document`).

Also could you please [sign our CLA](https://github.com/aseprite/opensource/blob/master/sign-cla.md#sign-the-cla)?

**Comment by zed9h** [comment]
@dacap, maybe It would be cleaner if I created a helper to decode/encode the PNG for both file formats, but first I will work on the Pixly format save, which will add more value upfront.

PS: I will read the CLA.

**Comment by dacap** [comment]
I'll take a look at this on next Tuesday. At the moment I recommend you to use `load_document()` if it's possible, so we can avoid modifying the actual png encoder/decoder (the code review from my side would be too much work if you refactor the code).

You can use `Document* tmpDoc = load_document(nullptr, pngfilename)` to "subload" the png file, and `save_document(nullptr, tmpDoc)` to save the png file.

**Comment by zed9h** [comment]
I signed the CLA and finished the Pixly file save; the code ended up much simpler than the original PNG format, because it only handle RGBA, take a look if you think it worth the generalization of the codec. For me it was easier to handle a standard lib code than aseprite internal structure, even when I found the latter pretty clear and organized.

**Comment by dacap** [comment]
@zed9h Remember that there is [an email required](https://github.com/aseprite/opensource/blob/master/sign-cla.md#sending-the-email) in the CLA signing process too.

File: src/app/file/pixly_format.cpp [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+  for (y = 0; y < height; y++) {
+    rows_pointer[y] = (png_bytep)png_malloc(png_ptr, png_get_rowbytes(png_ptr, info_ptr));
+    fop->setProgress(0.1 * (double)(y+1) / (double)height);
+  }	
+
+  fprintf(xml_fp, 
+    "<PixlyAnimation version=\"1.5\">\n"
+    "\t<Info "
+    "sheetWidth=\"%d\" sheetHeight=\"%d\" "
+    "totalCollumns=\"%d\" totalRows=\"%d\" "
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: Is this `totalCollumns` or `totalColumns`?

File: src/app/file/pixly_format.cpp [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+    TiXmlElement* xmlFrames = xmlAnim->FirstChild("Frames")->ToElement();
+    int imageCount = strtol(xmlFrames->Attribute("length"), NULL, 10);
+    int frameCount = imageCount / layerCount;
+    sprite->setTotalFrames(frame_t(frameCount));
+
+    for(int i=0; i<layerCount; i++) {
+      sprite->folder()->addLayer(new LayerImage(sprite));
+    }
+
+    int *visible = (int*)malloc(layerCount * sizeof(int));
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: `visible` can be a `std::vector`, e.g.

    std::vector<int> visible(layerCount, 0);

File: src/app/file/pixly_format.cpp [hunk 3]
```diff
@@ -0,0 +1,487 @@
+// Aseprite
+// Copyright (C) 2001-2016  David Capello
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: You can add a new Copyright (C) 2016 [your name] here.

File: src/app/file/pixly_format.cpp [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        break;
+    }
+  }
+
+  {
+    XmlDocumentRef doc = open_xml(fop->filename());
+    TiXmlHandle xml(doc.get());
+    fop->setProgress(0.75);
+
+    TiXmlElement* xmlAnim = xml.FirstChild("PixlyAnimation").ToElement();
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: What happen if `PixlyAnimation` element does not exist?

File: src/app/file/pixly_format.cpp [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+
+    TiXmlElement* xmlInfo = xmlAnim->FirstChild("Info")->ToElement();
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: Same here, what happen if `Info` element does not exist? (Or the attributes in the following lines, `layerCount`, `frameWidth`, `frameHeight`?)

File: src/app/file/pixly_format.cpp [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+
+    int layerCount  = strtol(xmlInfo->Attribute("layerCount"), NULL, 10);
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: You should use `base::convert_to<int>(...)` to convert from `std::string`/`const char*` to `int`

File: src/app/file/pixly_format.cpp [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+      Layer *layer = sprite->indexToLayer(layer_index);
+
+      const char * duration_str = xmlFrame->Attribute("duration");
+      if(duration_str) {
+        sprite->setFrameDuration(frame, strtol(duration_str, NULL, 10));
+      }
+
+      const char * visible_str = xmlFrame->Attribute("visible");
+      if(visible_str) {
+        visible[(int)layer_index] += std::string(visible_str) == "true";
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: Add `(`and `)` in the right-side expression:

    visible[(int)layer_index] += (std::string(visible_str) == "true");

**Comment by dacap** [comment]
As a side note, I'll need several `.anim` files to test these routines (original files generated with Pixly), because I don't have an Android device and cannot install Pixly. You can send me some test files at support@aseprite.org (I'll add them into an internal test suite I use for decoders/encoders)

**Comment by zed9h** [comment]
@dacap Great points, I placed many more check()s and improved the error messages. I will send some good files and some bad ones to test the exceptions.

File: src/app/file/pixly_format.cpp [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+  png_destroy_read_struct(&png_ptr, &info_ptr, NULL);
+  return success;
+}
+
+#ifdef ENABLE_SAVE
+bool PixlyFormat::onSave(FileOp* fop)
+{
+  const Sprite* sprite = fop->document()->sprite();
+
+  if(sprite->pixelFormat() != IMAGE_RGB || !fop->document()->sprite()->needAlpha()) {
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: There is no need to check `needAlpha`. Also, you can remove the `pixelFormat()` check if you remove the `FILE_SUPPORT_GRAY `, `FILE_SUPPORT_GRAYA`, and `FILE_SUPPORT_INDEXED` from `onGetFlags()` (and you should remove those constants from `onGetFlags()` as this format doesn't support Grayscale, Indexed)

File: src/app/file/pixly_format.cpp [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+      FILE_SUPPORT_LOAD |
+      FILE_SUPPORT_SAVE |
+      FILE_SUPPORT_RGB |
+      FILE_SUPPORT_RGBA |
+      FILE_SUPPORT_GRAY |
+      FILE_SUPPORT_GRAYA |
+      FILE_SUPPORT_INDEXED |
+      FILE_SUPPORT_LAYERS |
+      FILE_SUPPORT_FRAMES |
+      FILE_SUPPORT_PALETTE_WITH_ALPHA;
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: Remove `FILE_SUPPORT_PALETTE_WITH_ALPHA` too as this format doesn't support indexed images with color palettes (with alpha or without alpha).

File: src/app/file/pixly_format.cpp [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+#include "app/ini_file.h"
+#include "app/xml_document.h"
+#include "base/file_handle.h"
+#include "base/convert_to.h"
+#include "base/path.h"
+#include "doc/doc.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: As a minor detail, you can `#include <cstdio>`, `<cstdlib>`, and `<cmath>` here.

**Comment by zed9h** [comment]
If I remove the FILE_SUPPORT_PALETTE_WITH_ALPHA and the auto-generated palette has some alpha on it, there will be a warning every time the file is saved, so I think I will need to keep this flag, even if the palette will not be saved. I also added FILE_SUPPORT_BIG_PALETTES for the same reason. (Not really sure if this is the right way to go.)

I cleaned up some includes that seemed unnecessary.

**Comment by dacap** [comment]
Mhh, I think it's a bug in the code that check conditions in `file.cpp`. I'll be working in a fix for this so `FILE*PALETTE*` flags are checked only for Indexed images. (But I've to think more about those situations.)

The `FILE_SUPPORT_BIG_PALETTES` is in case the file format can save a color palette with more than 256 colors.

You can leave the flags and I'll then modify when the fix is done. Don't worry about that.

File: src/app/file/pixly_format.cpp [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+    int frameWidth  = check_number<int>(xmlInfo->Attribute("frameWidth"));
+    int frameHeight = check_number<int>(xmlInfo->Attribute("frameHeight"));
+
+    UniquePtr<Sprite> sprite(new Sprite(IMAGE_RGB, frameWidth, frameHeight, 0));
+
+    TiXmlElement* xmlFrames = check(xmlAnim->FirstChild("Frames"))->ToElement();
+    int imageCount = check_number<int>(xmlFrames->Attribute("length"));
+
+    if(layerCount <= 0 || imageCount <= 0) {
+      throw Exception("No cels found");
```
Comment (dacap) [inline src/app/file/pixly_format.cpp]: I see that these `throw` can generate some leaks (mainly the png structures that are not freed). Anyway, don't worry, I'll integrate this with the png encoder/decoder later. I'll merge this code as it is.

Comment (dacap) [inline src/app/file/pixly_format.cpp]: Oh sorry, I've just saw the try{}catch block to avoid leaks.

**Comment by dacap** [comment]
Finally I got a crash [saving this file](https://github.com/aseprite/aseprite/files/346562/crash.ase.zip):

screen shot 2016-07-04 at 12 34 09 pm

Could you please check this issue?

**Comment by zed9h** [comment]
As I suspected when I saw the line of the error, the cel image can be smaller than the sprite size, I will compensate for that.

**Comment by dacap** [comment]
Returning to this one: I'll finally remove `FILE_SUPPORT_BIG_PALETTES` and `FILE_SUPPORT_PALETTE_WITH_ALPHA` flags from `.anim` format as this format doesn't support these features (and the user must be warned that the palette will be lost). I've added the new issue #1182 how to fix the issues commented here.

**Comment by zed9h** [comment]
Of course, if the palette is user created there should be a warning.

@dacap BTW, I didn't explain it properly here, the "collums" (sic) are on the official Pixly format, nothing I could do about it.

**Comment by dacap** [comment]
> BTW, I didn't explain it properly here, the "collums" (sic) are on the official Pixly format, nothing I could do about it

👍   yeah, that's what I thought

**Pull request merged**